csharp/extractor/Semmle.Extraction.CSharp/Populators/CompilationUnit.cs

```diff
@@ -58,12 +59,15 @@ public override void VisitAttributeList(AttributeListSyntax node)
                 return;
 
             var outputAssembly = Assembly.CreateOutputAssembly(cx);
-            var attributeDatas = cx.Compilation.Assembly.GetAttributes().ToList();
-            attributeDatas.AddRange(cx.Compilation.Assembly.Modules.SelectMany(m => m.GetAttributes()));
+            var attributeLookup = new Dictionary<SyntaxNode, AttributeData>();
+            foreach (var attributeData in cx.Compilation.Assembly.GetAttributes().Concat(cx.Compilation.Assembly.Modules.SelectMany(m => m.GetAttributes())))
+            {
+                if (attributeData.ApplicationSyntaxReference?.GetSyntax() is SyntaxNode syntax)
+                    attributeLookup.Add(syntax, attributeData);
+            }
```

Comment:
This could be further optimized by making `attributeLookup` a field and only collecting this data once for the compilation, and not every time an attribute list is visited. It would be a bit more complex, because we would need locking around it. I can add this change...

Reply:
Good idea, I'll fix.